File: scripts/cohort-csv.ts
import { readFileSync } from "fs";
import path from "path";
import { parse } from "csv-parse/sync";
import type { Site, SiteTier } from "../lib/types";

// Reader for data/cohort.csv, the canonical 28-site cohort. Validates on the way in so CSV
// drift surfaces here with a row number instead of as a Postgres CHECK violation.
//
// Used by the seed script and Lane 1. Lane 2 reads the same file through the Python
// mirror, scripts/cohort_csv.py — keep the two readers in sync.

const TIERS: readonly SiteTier[] = [
  "anchor",
  "middle",
  "government",
  "small_business",
  "off_diagonal",
  "blocker",
];

export const COHORT_CSV_PATH = path.join(process.cwd(), "data", "cohort.csv");

export function readCohort(csvPath: string = COHORT_CSV_PATH): Site[] {
  const records = parse(readFileSync(csvPath, "utf8"), {
    columns: true,
    skip_empty_lines: true,
    trim: true,
  }) as Record<string, string>[];

  const file = path.basename(csvPath);
  const seen = new Set<string>();

  return records.map((row, i) => {
    const where = `${file} row ${i + 2}`; // +2: header plus 1-based rows

    const site_id = required(row, "site_id", where);
    if (seen.has(site_id)) throw new Error(`${where}: duplicate site_id "${site_id}"`);
    seen.add(site_id);

    const tier = required(row, "tier", where);
    if (!TIERS.includes(tier as SiteTier)) {
      throw new Error(`${where}: tier "${tier}" is not one of ${TIERS.join(", ")}`);
    }

    return {
      site_id,
      name: required(row, "name", where),
      tier: tier as SiteTier,
      start_url: required(row, "start_url", where),
      question: required(row, "question", where),
      answer_substring: required(row, "answer_substring", where),
      match_rule: row.match_rule ?? "",
      flag: row.flag ?? "",
      answer_note: row.answer_note ?? "",
    };
  });
}

function required(row: Record<string, string>, column: string, where: string): string {
  const value = (row[column] ?? "").trim();
  if (!value) throw new Error(`${where}: missing required column "${column}"`);
  return value;
}

File: scripts/cohort_csv.py
"""
Python mirror of scripts/cohort-csv.ts — reads data/cohort.csv, the canonical 28-site
cohort, validating on the way in so CSV drift surfaces with a row number. Keep the two
readers in sync; the TS one is used by the seed and Lane 1, this one by Lane 2 and the
scoring tests.

AGENTRANK_COHORT_CSV points the lanes at a different pre-registered row set without
touching the frozen loop or the canonical cohort. It exists for the authored Goodhart
exhibit (data/exhibit-cohort.csv, docs/EXHIBIT.md), which is measured by the same lanes
into its own batch_label and is never part of the 28-site cohort. Unset, the default is
byte-for-byte the previous behaviour, and a test pins that.
"""

from __future__ import annotations

import csv
import os
from pathlib import Path
from typing import Dict, List

TIERS = {"anchor", "middle", "government", "small_business", "off_diagonal", "blocker"}
REQUIRED = ("site_id", "name", "tier", "start_url", "question", "answer_substring")

REPO_ROOT = Path(__file__).resolve().parent.parent
DEFAULT_COHORT_CSV_PATH = REPO_ROOT / "data" / "cohort.csv"


def default_csv_path() -> Path:
    """The row set the lanes read: data/cohort.csv unless AGENTRANK_COHORT_CSV overrides it."""
    override = os.environ.get("AGENTRANK_COHORT_CSV")
    return Path(override) if override else DEFAULT_COHORT_CSV_PATH


COHORT_CSV_PATH = DEFAULT_COHORT_CSV_PATH


def read_cohort(csv_path: Path | None = None) -> List[Dict[str, str]]:
    csv_path = csv_path if csv_path is not None else default_csv_path()
    sites: List[Dict[str, str]] = []
    seen: set[str] = set()

    with open(csv_path, newline="", encoding="utf-8") as f:
        for i, row in enumerate(csv.DictReader(f)):
            where = f"{Path(csv_path).name} row {i + 2}"  # +2: header plus 1-based rows
            site = {k: (v or "").strip() for k, v in row.items() if k is not None}

            for col in REQUIRED:
                if not site.get(col):
                    raise ValueError(f'{where}: missing required column "{col}"')
            if site["site_id"] in seen:
                raise ValueError(f'{where}: duplicate site_id "{site["site_id"]}"')
            seen.add(site["site_id"])
            if site["tier"] not in TIERS:
                raise ValueError(
                    f'{where}: tier "{site["tier"]}" is not one of {", ".join(sorted(TIERS))}'
                )

            site.setdefault("match_rule", "")
            site.setdefault("flag", "")
            site.setdefault("answer_note", "")
            sites.append(site)

    if not sites:
        raise ValueError(f"{csv_path} has no rows")
    return sites
